includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-update.php

```diff
@@ -0,0 +1,271 @@
+
+			$subscription = $this->get_subscription_to_migrate( $subscription_id );
+			$source_id    = $subscription->get_meta( self::SOURCE_ID_META_KEY );
+			$user_id      = $subscription->get_user_id();
+
+			// Try to create an update SEPA gateway token if none exists.
+			// We don't need this to update the subscription, but creating one for consistency.
+			// It could be confusing for a merchant to see the subscription renewing but no saved token in the store.
+			$this->maybe_create_updated_sepa_token_by_source_id( $source_id, $user_id );
+
+			// Update the subscription with the updated SEPA gateway ID.
+			$this->set_subscription_updated_payment_method( $subscription );
+
+			$this->log( sprintf( 'Successful migration of subscription #%1$d.', $subscription_id ) );
+		} catch ( \Exception $e ) {
+			$this->log( $e->getMessage() );
+		}
+	}
+
+	/**
+	 * Gets the subscription to update.
+	 *
+	 * Only allows migration if:
+	 * - The Legacy experience is disabled
+	 * - The WooCommerce Subscription extension is active
+	 * - The subscription ID is a valid subscription
+	 *
+	 * @param int $subscription_id The ID of the subscription to migrate.
+	 * @return WC_Subscription The Subscription object for which its token must be updated.
+	 * @throws \Exception Skip the migration if the request is invalid.
+	 */
+	private function get_subscription_to_migrate( $subscription_id ) {
+		if ( ! WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. The Legacy experience is enabled.', $subscription_id ) );
+		}
+
+		if ( ! class_exists( 'WC_Subscriptions' ) ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. The WooCommerce Subscriptions extension is not active.', $subscription_id ) );
+		}
+
+		$subscription = wcs_get_subscription( $subscription_id );
+
+		if ( ! $subscription ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. Subscription not found.', $subscription_id ) );
+		}
+
+		return $subscription;
+	}
+
+	/**
+	 * Returns an updated token to be used for the subscription, given the source ID.
+	 *
+	 * If no updated token is found, we create a new one based on the legacy one.
+	 *
+	 * @param string  $source_id The Source or Payment Method ID associated with the subscription.
+	 * @param integer $user_id   The WordPress User ID to whom the subscription belongs.
+	 */
+	private function maybe_create_updated_sepa_token_by_source_id( string $source_id, int $user_id ) {
+
+		// Retrieve the updated SEPA tokens for the user.
+		$replacement_token = $this->get_customer_token_by_source_id( $source_id, $user_id, $this->updated_sepa_gateway_id );
+
+		// If no updated SEPA token was found, create a new one based on the source ID.
+		if ( ! $replacement_token ) {
+			$replacement_token = $this->create_updated_sepa_token( $source_id, $user_id );
+		}
```

Comment:
I might be confused here but this function reads as:

1. Find the token for the `$source_id`. 
2. If not token was found, create an updated one. 

But if I look at `create_updated_sepa_token()` it will first check if the token exists, and if it doesn't it will exit. 

Combined, doesn't that mean that if the legacy token doesn't exist, we call the create one, but it will skip because a legacy one doesn't exist?

Reply:
I was confused. :)

`$replacement_token = $this->create_updated_sepa_token( ... );` here tries to find the token using the **new gateway ID** (`stripe_sepa_debit`), if it doesn't find a saved token with the new gateway ID and a matching `pm_` it will create a new one. 

`create_updated_sepa_token()` will then attempt to find an existing token using the old ID (`stripe_sepa`)

ie if there's no token for `stripe_sepa_debit` create one but only if there's one with the `stripe_sepa`.